### PR TITLE
[gst-va] add vp9 encode test implementation

### PR DESCRIPTION
**Pull request by FocusLuo**
gst-va add vp9 encode test implementation

### CODE REVIEW

File: test/gst-va/encode/vp9.py [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+      loopshp   = loopshp,
+      qp        = qp,
+      quality   = quality,
+      rcmode    = "cqp",
+    )
+
+  @slash.parametrize(*gen_vp9_cqp_parameters(spec))
+  def test(self, case, ipmode, qp, quality, refmode, looplvl, loopshp):
+    self.init(spec, case, ipmode, qp, quality, refmode, looplvl, loopshp)
+    self.encode()
```
Comment (uartie) [inline test/gst-va/encode/vp9.py:66]: `vavp9enc` does not have `refmode` property... please use `@parametrize_with_unused`, then remove that parameter from the init function.

File: test/gst-va/encode/vp9.py [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+      loopshp   = loopshp,
+      maxrate   = bitrate,
+      minrate   = bitrate,
+      rcmode    = "cbr",
+    )
+ 
+  @slash.parametrize(*gen_vp9_cbr_parameters(spec))
+  def test(self, case, gop, bitrate, fps, refmode, looplvl, loopshp):
+    self.init(spec, case, gop, bitrate, fps, refmode, looplvl, loopshp)
+    self.encode()
```
Comment (uartie) [inline test/gst-va/encode/vp9.py:87]: `vavp9enc` does not have `refmode` property... please use `@parametrize_with_unused`, then remove that parameter from the init function.

File: test/gst-va/encode/vp9.py [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+      maxrate   = bitrate,
+      minrate   = bitrate,
+      quality   = quality,
+      rcmode    = "vbr",
+    )
+  
+  @slash.parametrize(*gen_vp9_vbr_parameters(spec))
+  def test(self, case, gop, bitrate, fps, refmode, quality, looplvl, loopshp):
+    self.init(spec, case, gop, bitrate, fps, refmode, quality, looplvl, loopshp)
+    self.encode()
```
Comment (uartie) [inline test/gst-va/encode/vp9.py:109]: `vavp9enc` does not have `refmode` property... please use `@parametrize_with_unused`, then remove that parameter from the init function.

File: test/gst-va/encode/vp9.py [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+      looplvl   = looplvl,
+      loopshp   = loopshp,
+      quality   = quality,
+      rcmode    = "cqp",
+    )
+
+  @parametrize_with_unused(*gen_vp9_cqp_lp_parameters(spec), ['slices'])
+  def test(self, case, ipmode, qp, quality, slices, refmode, looplvl, loopshp):
+    self.init(spec, case, ipmode, qp, quality, refmode, looplvl, loopshp)
+    self.encode()
```
Comment (uartie) [inline test/gst-va/encode/vp9.py:126]: `refmode` unused, too.  Also, remove that parameter from the init function.

File: test/gst-va/encode/vp9.py [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+      loopshp   = loopshp,
+      maxrate   = bitrate,
+      minrate   = bitrate,
+      rcmode    = "cbr",
+    )
+  
+  @parametrize_with_unused(*gen_vp9_cbr_lp_parameters(spec), ['slices'])
+  def test(self, case, gop, bitrate, fps, slices, refmode, looplvl, loopshp):
+    self.init(spec, case, gop, bitrate, fps, refmode, looplvl, loopshp)
+    self.encode()
```
Comment (uartie) [inline test/gst-va/encode/vp9.py:147]: `refmode` unused, too.  Also, remove that parameter from the init function.

File: test/gst-va/encode/vp9.py [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+      maxrate   = bitrate,
+      minrate   = bitrate,
+      quality   = quality,
+      rcmode    = "vbr",
+    )
+  
+  @parametrize_with_unused(*gen_vp9_vbr_lp_parameters(spec), ['slices'])
+  def test(self, case, gop, bitrate, fps, quality, slices, refmode, looplvl, loopshp):
+    self.init(spec, case, gop, bitrate, fps, quality, refmode, looplvl, loopshp)
+    self.encode()
```
Comment (uartie) [inline test/gst-va/encode/vp9.py:169]: `refmode` unused, too.  Also, remove that parameter from the init function.

File: test/gst-va/encode/vp9.py [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+    vars(self).update(tspec[case].copy())
+    vars(self).update(
+      bitrate   = bitrate,
+      case      = case,
+      fps       = fps,
+      frames    = vars(self).get("brframes", self.frames),
+      gop       = gop,
+      looplvl   = looplvl,
+      loopshp   = loopshp,
+      maxrate   = bitrate,
```
Comment (uartie) [inline test/gst-va/encode/vp9.py:160]: `maxrate = bitrate * 2` for vbr

**Review by uartie — CHANGES_REQUESTED** [review]

**Comment by uartie** [comment]
It looks like the `qp` property for vp9 is not converted correctly in `lib/gstreamer/va/encoder.py`.  Also, not sure why "vp8" is checked, too, in the `qp` property since gst-va doesn't provide an encoder for vp8.  Please fix.

File: test/gst-va/encode/vp9.py (hunk 7 above)
Comment (FocusLuo) [reply test/gst-va/encode/vp9.py:160]: updated

File: test/gst-va/encode/vp9.py (hunk 6 above)
Comment (FocusLuo) [reply test/gst-va/encode/vp9.py:169]: updated

File: test/gst-va/encode/vp9.py (hunk 5 above)
Comment (FocusLuo) [reply test/gst-va/encode/vp9.py:147]: updated

File: test/gst-va/encode/vp9.py [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
```
[the remark this answers is not in the document]
Comment (FocusLuo) [reply test/gst-va/encode/vp9.py:100]: updated

File: test/gst-va/encode/vp9.py [hunk 9]
```diff
@@ -0,0 +1,169 @@
...
+    return "webm"
+
+@slash.requires(*platform.have_caps("encode", "vp9_8"))
+class VP9EncoderTest(VP9EncoderBaseTest):
+  def before(self):
+    super().before()
+    vars(self).update(
+      caps          = platform.get_caps("encode", "vp9_8"),
+      lowpower      = False,
+      gstencoder    = "vavp9lpenc",
```
[the remark this answers is not in the document]
Comment (FocusLuo) [reply test/gst-va/encode/vp9.py:36]: updated

File: test/gst-va/encode/vp9.py [hunk 10]
```diff
@@ -0,0 +1,169 @@
...
+      gstparser     = "vp9parse",
+      gstmuxer      = "matroskamux",
+      gstdemuxer    = "matroskademux",
+    )
+
+  def get_file_ext(self):
+    return "webm"
+
+@slash.requires(*platform.have_caps("encode", "vp9_8"))
+class VP9EncoderTest(VP9EncoderBaseTest):
```
[the remark this answers is not in the document]
Comment (FocusLuo) [reply test/gst-va/encode/vp9.py:30]: thanks capture, updated

**Review by uartie — CHANGES_REQUESTED** [review]

File: test/gst-va/encode/vp9.py [hunk 11]
```diff
@@ -0,0 +1,171 @@
...
+      gop       = gop,
+      looplvl   = looplvl,
+      loopshp   = loopshp,
+      maxrate   = bitrate,
+      minrate   = bitrate,
+      rcmode    = "cbr",
+    )
+
+  @parametrize_with_unused(*gen_vp9_cbr_parameters(spec), ['slices', 'refmode'])
+  def test(self, case, gop, bitrate, fps, refmode, looplvl, loopshp):
```
[the remark this answers is not in the document]
Comment (FocusLuo) [reply test/gst-va/encode/vp9.py]: removed slices done

File: test/gst-va/encode/vp9.py [hunk 12]
```diff
@@ -0,0 +1,171 @@
...
+      gop       = 30 if ipmode != 0 else 1,
+      looplvl   = looplvl,
+      loopshp   = loopshp,
+      qp        = qp,
+      quality   = quality,
+      rcmode    = "cqp",
+    )
+
+  @parametrize_with_unused(*gen_vp9_cqp_parameters(spec), ['slices', 'refmode'])
+  def test(self, case, ipmode, qp, quality, refmode, looplvl, loopshp):
```
[the remark this answers is not in the document]
Comment (FocusLuo) [reply test/gst-va/encode/vp9.py]: Thanks Artie, Yes, let me remove the VME relative part
we will not support gst-va on i965, so let us remove vme vp9 encode. If need in future, then to submit new patch.

File: test/gst-va/encode/vp9.py [hunk 13]
```diff
@@ -0,0 +1,171 @@
...
+      looplvl   = looplvl,
+      loopshp   = loopshp,
+      maxrate   = bitrate * 2,
+      minrate   = bitrate,
+      quality   = quality,
+      rcmode    = "vbr",
+    )
+
+  @parametrize_with_unused(*gen_vp9_vbr_parameters(spec), ['slices', 'refmode'])
+  def test(self, case, gop, bitrate, fps, refmode, quality, looplvl, loopshp):
```
[the remark this answers is not in the document]
Comment (FocusLuo) [reply test/gst-va/encode/vp9.py]: removed slices

File: test/gst-va/encode/vp9.py [hunk 14]
```diff
@@ -0,0 +1,99 @@
...
+      case      = case,
+      gop       = 30 if ipmode != 0 else 1,
+      looplvl   = looplvl,
+      loopshp   = loopshp,
+      quality   = quality,
+      rcmode    = "cqp",
+      qp        = qp,
+    )
+
+  @parametrize_with_unused(*gen_vp9_cqp_lp_parameters(spec), ['refmode'])
```
Comment (uartie) [inline test/gst-va/encode/vp9.py:53]: `slices` should not have been removed here (only for the non-lp tests).  It is in the test parameters for the lp tests but it is not used.  Thus, it needs to be announced here.

File: test/gst-va/encode/vp9.py [hunk 15]
```diff
@@ -0,0 +1,99 @@
...
+      frames    = vars(self).get("brframes", self.frames),
+      gop       = gop,
+      looplvl   = looplvl,
+      loopshp   = loopshp,
+      maxrate   = bitrate,
+      minrate   = bitrate,
+      rcmode    = "cbr",
+    )
+  
+  @parametrize_with_unused(*gen_vp9_cbr_lp_parameters(spec), ['refmode'])
```
Comment (uartie) [inline test/gst-va/encode/vp9.py:74]: `slices` needs added to unused.

File: test/gst-va/encode/vp9.py [hunk 16]
```diff
@@ -0,0 +1,99 @@
...
+      gop       = gop,
+      looplvl   = looplvl,
+      loopshp   = loopshp,
+      maxrate   = bitrate * 2,
+      minrate   = bitrate,
+      quality   = quality,
+      rcmode    = "vbr",
+    )
+  
+  @parametrize_with_unused(*gen_vp9_vbr_lp_parameters(spec), ['refmode'])
```
Comment (uartie) [inline test/gst-va/encode/vp9.py:96]: `slices` needs added to unused.

**Review by uartie — CHANGES_REQUESTED** [review]

File: test/gst-va/encode/vp9.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (uartie) [reply test/gst-va/encode/vp9.py]: Don't remove from LP tests, only the non-LP needed removed.

File: test/gst-va/encode/vp9.py (hunk 14 above)
Comment (FocusLuo) [reply test/gst-va/encode/vp9.py]: right, my mistake, added slices back for lp mode

**Pull request merged**